lib/src/modules/task.cc

```diff
@@ -128,11 +147,150 @@ static TaskCommand getTaskCommand(const fs::path &task_executable)
+    } else if (err_code ==  boost_error::file_exists) {
+        return Module::ProcessingError(lth_loc::format("A file matching the name of the provided sha already exists"));
+    } else {
+        return Module::ProcessingError(e.what());
+    }
+}
+
+// Take out common permissions.
+static fs::perms TASK_FILE_PERMS = fs::owner_read | fs::owner_write | fs::group_read;
+
+// Creates the <task_cache_dir>/<sha256> directory for a single
+// task, ensuring that its permissions are readable by
+// the PXP agent owner/group (for unix OSes), writable for the PXP agent owner,
+// and executable by both PXP agent owner and group. Returns the path to this directory.
+// Note that the last modified time of the directory is updated, and that this routine
+// will not fail if the directory already exists.
+static fs::path createCacheDir(const fs::path& task_cache_dir, const std::string& sha256) {
+    auto cache_dir = task_cache_dir / sha256;
+    fs::create_directory(cache_dir);
+    fs::last_write_time(cache_dir, time(nullptr));
+    fs::permissions(cache_dir, TASK_FILE_PERMS | fs::owner_exe | fs::group_exe);
+    return cache_dir;
+}
+
+// Computes the sha256 of the file denoted by path. Assumes that
+// the file designated by "path" exists.
+static std::string calculateSha256(const std::string& path) {
+    const std::streamsize CHUNK_SIZE = 2 << 14;
+    char buffer[CHUNK_SIZE];
+    unsigned char md_value[EVP_MAX_MD_SIZE];
+    unsigned int md_len;
+
+    auto md = EVP_sha256();
+    auto mdctx = EVP_MD_CTX_create();
+    EVP_DigestInit_ex(mdctx, md, nullptr);
+    boost::nowide::ifstream ifs(path);
+    while (ifs.read(buffer, CHUNK_SIZE)) {
+      EVP_DigestUpdate(mdctx, buffer, CHUNK_SIZE);
+    }
+    if (!ifs.eof()) {
+      throw Module::ProcessingError(lth_loc::format("Error while reading {1}", path));
+    }
+    EVP_DigestUpdate(mdctx, buffer, ifs.gcount());
+    EVP_DigestFinal_ex(mdctx, md_value, &md_len);
+    EVP_MD_CTX_destroy(mdctx);
+
+    char md_value_ascii[2*EVP_MAX_MD_SIZE+1] = {'\0'};
+    for (unsigned int i = 0; i < md_len; ++i) {
+      snprintf(md_value_ascii+2*i, 3, "%02x", md_value[i]);
+    }
+    return std::string(md_value_ascii);
+}
+
+static std::string createUrlEndpoint(const lth_jc::JsonContainer& uri) {
+    std::string url = uri.get<std::string>("path");
+    auto params = uri.getWithDefault<lth_jc::JsonContainer>("params", lth_jc::JsonContainer());
+    if (params.empty()) {
+        return url;
+    }
+    auto curl_handle = lth_curl::curl_handle();
+    url += "?";
+    for (auto& key : params.keys()) {
+        auto escaped_key = std::string(
+           lth_curl::curl_escaped_string(curl_handle, key));
+        auto escaped_val = std::string(
+           lth_curl::curl_escaped_string(curl_handle, params.get<std::string>(key)));
+        url += escaped_key + "=" + escaped_val + "&";
+    }
+    return url;
+}
+
+// Downloads the file at the specified url into the provided path. If something goes 
```

Comment:
@MikaelSmith How is this for the documentation of the two temp-files case? I also refactored downloadTaskFile to remove the try catch block; the catch block for http_file_download_exception is in getCachedTask now.